vhost/src/vhost_user/frontend_req_handler.rs

```diff
@@ -395,6 +437,16 @@ mod tests {
             }
             Ok(1)
         }
+
+        fn shmem_map(&mut self, req: &VhostUserMMap, _fd: &dyn AsRawFd) -> HandlerResult<u64> {
+            assert_eq!({ req.shmid }, 0);
+            Ok(!self.shmem_mappings.insert((req.shm_offset, req.len)) as u64)
```

Comment:
i think it's better to remove the negation (!) and use a direct check to return meaningful values. e.g `if self.shmem_mappings.insert(req.shm_offset, req.len) { Ok(1) } else { Ok(0) }` and why the double braces? missing &?

Reply:
By braces do you mean `( )`? The `(req.shm_offset, req.len)` is a tuple.

In regards to the negation with cast instead of explicit `if`, I considered to write it like you did, I mainly wrote it like this, because it is symmetric with the existing `shared_object_add`, `shared_object_remove` above:
```Rust
fn shared_object_add(&mut self, uuid: &VhostUserSharedMsg) -> HandlerResult<u64> {
    Ok(!self.shared_objects.insert(uuid.uuid) as u64)
}
fn shared_object_remove(&mut self, uuid: &VhostUserSharedMsg) -> HandlerResult<u64> {
    Ok(!self.shared_objects.remove(&uuid.uuid) as u64)
}
```

I am not sure if I should change it.

Reply:
>  The (req.shm_offset, req.len) is a tuple.

ah right!

> because it is symmetric with the existing shared_object_add, shared_object_remove above:

doesn't necessarily mean it's the correct approach.

Reply:
The `{ }` braces are required here to force the value to be copied. Without the braces this fails to compile because  `assert_eq!` macro tries to take a reference to `req.shmid` which is not possible. This is because the struct `VhostUserMMap` is `repr(packed)` so it is `unsafe` to take references to its fields (they may not be aligned).